File: rust/rush-providers/src/auth/keycloak.rs
//! Keycloak token provider — mirrors hush-providers auth/keycloak.py.
//!
//! Fetches and caches OAuth2 tokens from a Keycloak endpoint.
//! Background refresh via tokio task (daemon-like, doesn't block shutdown).

use std::sync::Arc;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use tokio::sync::Mutex;

use crate::http::{get_client, ProviderError, ProviderResult};

/// Keycloak token provider configuration.
#[derive(Debug, Clone)]
pub struct KeycloakConfig {
    /// Token endpoint URL.
    pub url: String,
    /// Client name/ID.
    pub name: String,
    /// Client secret.
    pub secret: String,
    /// JSON path to extract token from response (default: "accessToken").
    pub token_path: String,
    /// JSON path to extract expiry from response (default: "expiresIn").
    pub expires_in_path: Option<String>,
    /// Background refresh interval in seconds (default: 3600).
    pub refresh_interval: f64,
    /// Seconds before expiry to proactively refresh (default: 300).
    pub refresh_buffer: f64,
}

impl Default for KeycloakConfig {
    fn default() -> Self {
        KeycloakConfig {
            url: String::new(),
            name: String::new(),
            secret: String::new(),
            token_path: "accessToken".to_string(),
            expires_in_path: Some("expiresIn".to_string()),
            refresh_interval: 3600.0,
            refresh_buffer: 300.0,
        }
    }
}

/// Thread-safe Keycloak token provider.
/// Caches tokens and refreshes proactively before expiry.
pub struct KeycloakTokenProvider {
    config: KeycloakConfig,
    state: Arc<Mutex<TokenState>>,
}

struct TokenState {
    token: Option<String>,
    expires_at: f64, // UNIX timestamp
}

impl KeycloakTokenProvider {
    pub fn new(config: KeycloakConfig) -> Self {
        KeycloakTokenProvider {
            config,
            state: Arc::new(Mutex::new(TokenState {
                token: None,
                expires_at: 0.0,
            })),
        }
    }

    /// Get a valid token — returns cached if not expired, fetches new otherwise.
    pub async fn get_token(&self) -> ProviderResult<String> {
        let mut state = self.state.lock().await;

        let now = current_timestamp();
        if let Some(ref token) = state.token {
            if now < state.expires_at {
                return Ok(token.clone());
            }
        }

        // Token missing or expired — fetch new
        let (token, expires_at) = self.fetch_token().await?;
        state.token = Some(token.clone());
        state.expires_at = expires_at;
        Ok(token)
    }

    /// Invalidate cached token (force refresh on next call).
    pub async fn invalidate(&self) {
        let mut state = self.state.lock().await;
        state.token = None;
        state.expires_at = 0.0;
    }

    /// Fetch a new token from the Keycloak endpoint.
    async fn fetch_token(&self) -> ProviderResult<(String, f64)> {
        let client = get_client();

        let body = serde_json::json!({
            "Name": &self.config.name,
            "Secret": &self.config.secret,
        });

        let resp = client
            .post(&self.config.url)
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await
            .map_err(|e| ProviderError {
                message: format!("Keycloak token fetch failed: {}", e),
                status_code: None,
                error_code: None,
            })?;

        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            return Err(ProviderError {
                message: format!("Keycloak error (HTTP {}): {}", status.as_u16(), body),
                status_code: Some(status.as_u16()),
                error_code: None,
            });
        }

        let data: serde_json::Value = resp.json().await.map_err(|e| ProviderError {
            message: format!("Failed to parse Keycloak response: {}", e),
            status_code: Some(status.as_u16()),
            error_code: None,
        })?;

        // Extract token using token_path
        let token = data
            .get(&self.config.token_path)
            .and_then(|v| v.as_str())
            .ok_or_else(|| ProviderError {
                message: format!(
                    "Keycloak response missing '{}' field",
                    self.config.token_path
                ),
                status_code: None,
                error_code: None,
            })?
            .to_string();

        // Extract expiry
        let now = current_timestamp();
        let expires_at = if let Some(ref expires_path) = self.config.expires_in_path {
            let expires_in = data
                .get(expires_path)
                .and_then(|v| v.as_f64())
                .unwrap_or(self.config.refresh_interval);
            now + expires_in - self.config.refresh_buffer
        } else {
            now + self.config.refresh_interval - self.config.refresh_buffer
        };

        Ok((token, expires_at))
    }

    /// Start a background refresh task (fire-and-forget).
    ///
    /// The task refreshes the token periodically. Runs until the provider
    /// is dropped (uses weak reference pattern via Arc).
    pub fn start_background_refresh(&self, runtime: &tokio::runtime::Runtime) {
        let config = self.config.clone();
        let state = Arc::clone(&self.state);
        let weak_state = Arc::downgrade(&self.state);

        runtime.spawn(async move {
            let interval = Duration::from_secs_f64(config.refresh_interval);
            loop {
                tokio::time::sleep(interval).await;

                // Stop if the provider has been dropped
                if weak_state.upgrade().is_none() {
                    break;
                }

                // Check if token needs refresh
                let needs_refresh = {
                    let s = state.lock().await;
                    let now = current_timestamp();
                    s.token.is_none() || now >= s.expires_at
                };

                if needs_refresh {
                    match fetch_keycloak_token(&config).await {
                        Ok((token, expires_at)) => {
                            let mut s = state.lock().await;
                            s.token = Some(token);
                            s.expires_at = expires_at;
                        }
                        Err(_) => {
                            // Log error but continue — next get_token() will retry
                        }
                    }
                }
            }
        });
    }
}

/// Standalone fetch function for background refresh task.
async fn fetch_keycloak_token(config: &KeycloakConfig) -> ProviderResult<(String, f64)> {
    let client = get_client();

    let body = serde_json::json!({
        "Name": &config.name,
        "Secret": &config.secret,
    });

    let resp = client
        .post(&config.url)
        .header("Content-Type", "application/json")
        .json(&body)
        .send()
        .await?;

    let status = resp.status();
    if !status.is_success() {
        let body = resp.text().await.unwrap_or_default();
        return Err(ProviderError {
            message: format!("Keycloak refresh error: {}", body),
            status_code: Some(status.as_u16()),
            error_code: None,
        });
    }

    let data: serde_json::Value = resp.json().await.map_err(|e| ProviderError {
        message: format!("Failed to parse Keycloak response: {}", e),
        status_code: Some(status.as_u16()),
        error_code: None,
    })?;

    let token = data
        .get(&config.token_path)
        .and_then(|v| v.as_str())
        .unwrap_or_default()
        .to_string();

    let now = current_timestamp();
    let expires_at = if let Some(ref expires_path) = config.expires_in_path {
        let expires_in = data
            .get(expires_path)
            .and_then(|v| v.as_f64())
            .unwrap_or(config.refresh_interval);
        now + expires_in - config.refresh_buffer
    } else {
        now + config.refresh_interval - config.refresh_buffer
    };

    Ok((token, expires_at))
}

fn current_timestamp() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or(Duration::ZERO)
        .as_secs_f64()
}

/// Deserialize KeycloakConfig from a serde_json::Value dict.
impl KeycloakConfig {
    pub fn from_json(val: &serde_json::Value) -> Option<Self> {
        let obj = val.as_object()?;
        Some(KeycloakConfig {
            url: obj.get("url")?.as_str()?.to_string(),
            name: obj.get("name")?.as_str()?.to_string(),
            secret: obj.get("secret")?.as_str()?.to_string(),
            token_path: obj
                .get("token_path")
                .and_then(|v| v.as_str())
                .unwrap_or("accessToken")
                .to_string(),
            expires_in_path: obj
                .get("expires_in_path")
                .and_then(|v| v.as_str())
                .map(|s| s.to_string()),
            refresh_interval: obj
                .get("refresh_interval")
                .and_then(|v| v.as_f64())
                .unwrap_or(3600.0),
            refresh_buffer: obj
                .get("refresh_buffer")
                .and_then(|v| v.as_f64())
                .unwrap_or(300.0),
        })
    }
}

File: python/hush-providers/hush/providers/auth/keycloak.py
"""Keycloak token provider with background refresh.

This module provides a thread-safe token provider that fetches OAuth tokens
from Keycloak endpoints with automatic background refresh.
"""

import atexit
import threading
import time
from typing import Optional

import httpx

from hush.core.loggings import LOGGER

from .config import KeycloakTokenConfig


class KeycloakTokenProvider:
    """Thread-safe Keycloak token provider with background refresh.

    Features:
    - Lazy token fetching on first access
    - Background daemon thread for proactive refresh
    - Thread-safe access via threading.Lock
    - Graceful shutdown via threading.Event
    - Error isolation (background errors logged, not raised)

    The background refresh runs as a daemon thread that:
    - Wakes up at refresh_interval or when token is near expiry
    - Does NOT interfere with main flow (errors are logged only)
    - Automatically stops when main program exits

    Example:
        config = KeycloakTokenConfig(
            url="https://identity.example.com/client/connect",
            name="my_app",
            secret="my_secret",
            token_path="accessToken"
        )
        provider = KeycloakTokenProvider(config)

        # Get token (fetches if needed, always returns valid token)
        token = provider.get_token()

        # Shutdown background refresh (optional, auto on exit)
        provider.shutdown()
    """

    # Track all providers for cleanup
    _instances: dict[str, "KeycloakTokenProvider"] = {}
    _instances_lock = threading.Lock()

    def __init__(self, config: KeycloakTokenConfig):
        """Initialize the token provider.

        Args:
            config: KeycloakTokenConfig with endpoint and credentials
        """
        self.config = config
        self._token: Optional[str] = None
        self._expires_at: float = 0
        self._lock = threading.Lock()
        self._shutdown_event = threading.Event()
        self._refresh_thread: Optional[threading.Thread] = None
        self._started = False

        # Register for cleanup
        with self._instances_lock:
            self._instances[config.name] = self

    def _fetch_token(self) -> str:
        """Fetch a new token from Keycloak endpoint.

        Returns:
            Access token string

        Raises:
            httpx.HTTPStatusError: If HTTP request fails
            ValueError: If token not found in response
        """
        try:
            with httpx.Client(verify=False, timeout=30.0) as client:
                response = client.post(
                    self.config.url,
                    json={"Name": self.config.name, "Secret": self.config.secret},
                    headers={"Content-Type": "application/json"},
                )
                response.raise_for_status()
                data = response.json()

                # Extract token
                token = data.get(self.config.token_path)
                if not token:
                    raise ValueError(f"Token not found at '{self.config.token_path}' in response")

                # Extract expiry
                expires_in = None
                if self.config.expires_in_path:
                    expires_in = data.get(self.config.expires_in_path)

                # Set expiry with buffer
                if expires_in:
                    self._expires_at = time.time() + expires_in - self.config.refresh_buffer
                else:
                    # Default 59 minutes if not provided
                    self._expires_at = time.time() + 3540

                self._token = token
                LOGGER.info(
                    "[KeycloakTokenProvider:%s] Token fetched, expires in %ds",
                    self.config.name,
                    expires_in or 3600,
                )
                return token

        except httpx.HTTPStatusError as e:
            LOGGER.error(
                "[KeycloakTokenProvider:%s] HTTP error: %d - %s",
                self.config.name,
                e.response.status_code,
                e.response.text,
            )
            raise
        except Exception as e:
            LOGGER.error(
                "[KeycloakTokenProvider:%s] Error fetching token: %s",
                self.config.name,
                e,
            )
            raise

    def _background_refresh_loop(self):
        """Background thread loop for proactive token refresh.

        This loop:
        1. Waits for refresh_interval or until shutdown
        2. Checks if token needs refresh (near expiry)
        3. Fetches new token if needed
        4. Logs errors but does NOT crash
        """
        LOGGER.info(
            "[KeycloakTokenProvider:%s] Background refresh started (interval=%ds)",
            self.config.name,
            self.config.refresh_interval,
        )

        while not self._shutdown_event.is_set():
            # Wait for next refresh interval or shutdown
            self._shutdown_event.wait(timeout=self.config.refresh_interval)

            if self._shutdown_event.is_set():
                break

            # Check if refresh is needed
            with self._lock:
                needs_refresh = not self._token or time.time() >= (
                    self._expires_at - self.config.refresh_buffer
                )

            if needs_refresh:
                try:
                    with self._lock:
                        self._fetch_token()
                    LOGGER.debug(
                        "[KeycloakTokenProvider:%s] Background refresh successful",
                        self.config.name,
                    )
                except Exception as e:
                    # Log but don't crash - on-demand fetch is fallback
                    LOGGER.warning(
                        "[KeycloakTokenProvider:%s] Background refresh failed: %s",
                        self.config.name,
                        e,
                    )

        LOGGER.info(
            "[KeycloakTokenProvider:%s] Background refresh stopped",
            self.config.name,
        )

    def _ensure_started(self):
        """Start background refresh thread if not already started."""
        if self._started:
            return

        with self._lock:
            if self._started:
                return

            # Clear shutdown event in case of restart after shutdown
            self._shutdown_event.clear()

            self._refresh_thread = threading.Thread(
                target=self._background_refresh_loop,
                name=f"KeycloakRefresh-{self.config.name}",
                daemon=True,  # Won't block program exit
            )
            self._refresh_thread.start()
            self._started = True

    def get_token(self) -> str:
        """Get a valid access token.

        Thread-safe. Fetches new token if:
        - No token cached
        - Token has expired

        Also starts background refresh thread on first call.

        Returns:
            Valid access token string

        Raises:
            Exception: If token fetch fails
        """
        # Start background refresh (idempotent)
        self._ensure_started()

        with self._lock:
            # Return cached token if valid
            if self._token and time.time() < self._expires_at:
                return self._token

            # Fetch new token
            return self._fetch_token()

    def invalidate(self):
        """Force token refresh on next get_token() call."""
        with self._lock:
            self._token = None
            self._expires_at = 0

    def shutdown(self):
        """Stop background refresh thread.

        Called automatically on program exit via atexit.
        """
        self._shutdown_event.set()
        if self._refresh_thread and self._refresh_thread.is_alive():
            self._refresh_thread.join(timeout=2.0)
        self._started = False

    @classmethod
    def shutdown_all(cls):
        """Shutdown all registered providers."""
        with cls._instances_lock:
            for provider in cls._instances.values():
                provider.shutdown()
            cls._instances.clear()


# Register cleanup on program exit
atexit.register(KeycloakTokenProvider.shutdown_all)
